clients/src/main/java/org/apache/kafka/clients/producer/internals/RecordAccumulator.java

```diff
@@ -70,6 +70,7 @@
     private final Time time;
     private final ConcurrentMap<TopicPartition, Deque<RecordBatch>> batches;
     private final IncompleteRecordBatches incomplete;
+    private final Set<TopicPartition> muted;
```

Comment:
It would be good to add a comment why this doesn't have to be thread-safe (unlike `incomplete` and `batches`, for example). In fact, it would be good to group the fields that are not thread-safe together (`drainIndex` seems to be the other one).